src/aws/flb_aws_util.c

```diff
@@ -507,3 +511,171 @@ int flb_read_file(const char *path, char **out_buf, size_t *out_size)
+        return NULL;
+    }
+
+    i = 0; 
+    while (*original_string) { 
+        if (strstr(original_string, current_word) == original_string) { 
+            strcpy(&result[i], new_word); 
+            i += new_word_len; 
+            original_string += old_word_len; 
+        } 
+        else
+            result[i++] = *original_string++; 
+    } 
+
+    result[i] = '\0'; 
+    return result; 
+} 
+
+/* Constructs S3 object key as per the format. */
+flb_sds_t get_s3_key(const char *format, time_t time, char *tag, char *tag_delimiter)
+{
+    int i = 0;
+    int ret = 0;
+    char *tag_token = NULL;
+    flb_sds_t tmp = NULL;
+    flb_sds_t buf = NULL;
+    flb_sds_t s3_key = NULL;
+    flb_sds_t tmp_key = NULL;
+    flb_sds_t tmp_tag = NULL;
+    struct tm *gmt = NULL;
+    
+    if (strlen(format) > S3_KEY_SIZE){
+        flb_warn("[s3_key] Object key length is longer than the 1024 character limit.");
+        return NULL;
+    }
+
+    tmp_tag = flb_sds_create_len(tag, strlen(tag) + 1);
+    if(!tmp_tag){
+        flb_errno();
+        return NULL;
+    }
+
+    s3_key = flb_sds_create_len(format, S3_KEY_SIZE);
+    if (!s3_key) {
+        flb_errno();
+        return NULL;
+    }
+
+    /* Check if delimiter(s) specifed exists in the tag. */
+    for (i = 0; i < strlen(tag_delimiter); i++){
+        if (strchr(tag, tag_delimiter[i])){
+            ret = 1;
+            break;
+        } 
+    }
+
+    if(ret == 0){
+        flb_warn("[s3_key] Invalid Tag delimiter: does not exist in tag.");
+        flb_sds_destroy(tmp_tag);
+        flb_sds_destroy(s3_key);
+        return NULL;
+    }  
+
+    /* Split the string on the delimiters */
+    tag_token = strtok(tmp_tag, tag_delimiter);
+    
+    /* Find all occurences of $TAG[*] and  
+     * replaces it with the right token from tag.
+     */
+    while(tag_token != NULL && i <= 10) {
+        buf = flb_sds_create_size(10);
+        tmp = flb_sds_printf(&buf, TAG_PART_DESCRIPTOR, i);
+
+        tmp_key = replace_uri_tokens(s3_key, tmp, tag_token);
+        if (!tmp_key) {
+            flb_errno();
+            return NULL;
+        }
+
+        if(strlen(tmp_key) > S3_KEY_SIZE){
+            flb_warn("[s3_key] Object key length is longer than the 1024 character limit.");
+            flb_sds_destroy(tmp);
+            flb_sds_destroy(tmp_tag);
+            flb_sds_destroy(s3_key);
+            flb_sds_destroy(tmp_key);
+            return NULL; 
+        }
+
+        flb_sds_destroy(tmp);
+        flb_sds_destroy(s3_key);
+        s3_key = tmp_key;
+
+        tag_token = strtok(NULL, tag_delimiter);
+        i++;
+    }
+    
+    tmp = flb_sds_create_size(10);
+    strncat(tmp, TAG_PART_DESCRIPTOR, 5);
+
+    if ((i <= 10 || tag_token != NULL) && strstr(s3_key, tmp)){
+        flb_warn("[s3_key] Invalid / Out of bounds tag part specified in the s3 key.");
+        flb_sds_destroy(tmp);
+        flb_sds_destroy(tmp_tag);
+        flb_sds_destroy(tmp_key);
+        return NULL;
+    }
```

Comment:
I think this should not be an error case. 

If user specifies $TAG[5] but the tag only has 4 parts, I think the code should continue working with just a warning. Users do not always carefully watch their tags- I wouldn't want Fluent Bit to fail to send logs just because of that.